File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int dp[502][502];

int solution(vector<vector<int>> arr) {
    int n = arr.size();

    for(int i = 0; i < n; i++){
        for(int j = 0; j <= i; j++){
          dp[i][j] = arr[i][j] + max(dp[i-1][j-1], dp[i-1][j]);
        }
    }
    
    int mx = -1;
    for(int i = 0; i < n; i++){
        mx = max(dp[n-1][i],mx);
    }

    return mx;
}

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.java
import java.io.*;
import java.util.*;

public class Main {

    static int arr[][] = new int[502][502];
    static int dp[][] = new int[502][502];

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int N = Integer.parseInt(br.readLine());
        for (int i = 1; i <= N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 1; j <= i; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        // 자신의 위쪽왼쪽, 위쪽 에서부터 누적합 중 큰 값 
        for (int i = 1; i <= N; i++) {
            for (int j = 1; j <= i; j++) {
                dp[i][j] = Math.max(dp[i - 1][j - 1], dp[i - 1][j]) + arr[i][j];
            }
        }

        int mx = 0;
        for (int i = 1; i <= N; i++) {
            mx = Math.max(mx, dp[N][i]);
        }
        
        System.out.println(mx);
    }


}